src/nf/nf_flatten_layer.f90

```diff
@@ -18,13 +18,20 @@ module nf_flatten_layer
     integer, allocatable :: input_shape(:)
     integer :: output_size
 
-    real, allocatable :: gradient(:,:,:)
+    real, allocatable :: gradient_2d(:,:)
+    real, allocatable :: gradient_3d(:,:,:)
```

Comment:
Yep, I thought about that but decided not to make the code even less SOLID

Reply:
But here we have a choice between SOLID and less boilerplate, I think I agree that the second one is better

Reply:
Yes, and, most importantly for me, this approach allows for a unified API (only one `flatten()` for the user).